src/test/java/com/balako/onlinebookstore/controller/BookControllerTest.java

```diff
@@ -0,0 +1,233 @@
+import com.balako.onlinebookstore.model.Category;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Set;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.MediaType;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class BookControllerTest {
+    protected static MockMvc mockMvc;
+    private static final CreateBookRequestDto REQUEST_DTO_1 =
+            new CreateBookRequestDto();
+    private static final CreateBookRequestDto REQUEST_DTO_2 =
+            new CreateBookRequestDto();
+    private static final Category VALID_CATEGORY_1 = new Category();
+    private static final Category VALID_CATEGORY_2 = new Category();
+    private static final BookDto VALID_BOOK_DTO_1 = new BookDto();
+    private static final BookDto VALID_BOOK_DTO_2 = new BookDto();
+    private static final BookSearchParametersDto SEARCH_PARAMETERS =
+            new BookSearchParametersDto(
+                    new String[] {"Title 1", "Title 2"},
+                    new String[] {"Author 1", "Author 2"},
+                    new String[] {"978-0545010221"});
+    private static final String ADMIN_USERNAME = "admin";
+    private static final String ADMIN_ROLE = "ADMIN";
+    private static final String USER_USERNAME = "user";
+    private static final String SQL_SCRIPT_BEFORE_METHOD =
+            "classpath:db/book/add-book-to-book-table"
+                    + "-and-category-to-category-table.sql";
+    private static final String SQL_SCRIPT_AFTER_METHOD =
+            "classpath:db/book/delete-from-book-and-category-table.sql";
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeAll
+    static void beforeAll(@Autowired WebApplicationContext webApplicationContext) {
+        mockMvc = MockMvcBuilders.webAppContextSetup(webApplicationContext)
+                .apply(springSecurity())
+                .build();
+
+        VALID_CATEGORY_1.setId(1L);
+        VALID_CATEGORY_1.setName("Category 1");
+        VALID_CATEGORY_1.setDescription("Category 1 description");
+
+        VALID_CATEGORY_2.setId(2L);
+        VALID_CATEGORY_2.setName("Category 2");
+        VALID_CATEGORY_2.setDescription("Category 2 description");
+
```

Comment:
You can use @Accessors(chain = true) under the model to write .set().set() thats will improve your code

Reply:
I know about this annotation, but it still in experimental package, so I choose a Jedi way.